Misleading error message when accessing a property in violation of noPropertyAccessFromIndexSignature

When the `noPropertyAccessFromIndexSignature` TypeScript compiler option is enabled and a template attempts a property access that fails this check, the error message is misleading.

Example:
```typescript

const record: Record<string, string> = {};

<template>
  {{record.foo}}
</template>
```

Error message:

```
error TS4111: Property 'foo' comes from an index signature, so it must be accessed with ['foo'].
```

It is not possible to write `{{record['foo']}}` in a handlebars template because this is not valid syntax, so the error message should be rewritten to suggest the `{{get}}` helper instead.

--- a/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
+++ b/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
@@ -880,4 +880,66 @@ describe('Language Server: Diagnostic Augmentation', () => {
       ]
     `);
   });
+
+  test('`noPropertyAccessFromIndexSignature` violation', () => {
+    project.updateTsconfig((tsconfig) => {
+      tsconfig.glint = { environment: ['ember-loose', 'ember-template-imports'] };
+      tsconfig.compilerOptions ??= {};
+      tsconfig.compilerOptions['noPropertyAccessFromIndexSignature'] = true;
+    });
+
+    project.write({
+      'index.gts': stripIndent`
+        declare const stringRecord: Record<string, string>;
+
+        stringRecord.fooBar;
+
+        <template>
+          {{stringRecord.fooBar}}          
+        </template>
+      `,
+    });
+
+    let server = project.startLanguageServer();
+    let diagnostics = server.getDiagnostics(project.fileURI('index.gts'));
+
+    expect(diagnostics).toMatchInlineSnapshot(`
+      [
+        {
+          "code": 4111,
+          "message": "Property 'fooBar' comes from an index signature, so it must be accessed with ['fooBar'].",
+          "range": {
+            "end": {
+              "character": 19,
+              "line": 2,
+            },
+            "start": {
+              "character": 13,
+              "line": 2,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(4111)",
+          "tags": [],
+        },
+        {
+          "code": 4111,
+          "message": "Property 'fooBar' comes from an index signature, so it must be accessed with {{get ... 'fooBar'}}.",
+          "range": {
+            "end": {
+              "character": 23,
+              "line": 5,
+            },
+            "start": {
+              "character": 17,
+              "line": 5,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(4111)",
+          "tags": [],
+        },
+      ]
+    `);
+  });
 });

--- a/packages/core/src/transform/diagnostics/augmentation.ts
+++ b/packages/core/src/transform/diagnostics/augmentation.ts
@@ -32,6 +32,7 @@ const diagnosticHandlers: Record<number, DiagnosticHandler | undefined> = {
   2554: noteNamedArgsAffectArity, // TS2554: Expected N arguments, but got M.
   2555: noteNamedArgsAffectArity, // TS2555: Expected at least N arguments, but got M.
   2769: checkResolveError, // TS2769: No overload matches this call.
+  4111: checkIndexAccessError, // TS4111: Property 'x' comes from an index signature, so it must be accessed with ['x'].
   7053: checkImplicitAnyError, // TS7053: Element implicitly has an 'any' type because expression of type '"X"' can't be used to index type 'Y'.
 };
 
@@ -208,6 +209,18 @@ function checkImplicitAnyError(
           `for this value; see the Template Registry page in the Glint documentation for more details.`
       );
     }
+  }
+}
+
+function checkIndexAccessError(diagnostic: Diagnostic, mapping: MappingTree): string | undefined {
+  if (mapping.sourceNode.type === 'Identifier') {
+    let messageText =
+      typeof diagnostic.messageText === 'string'
+        ? diagnostic.messageText
+        : diagnostic.messageText.messageText;
+
+    // "accessed with ['x']" => "accessed with {{get ... 'x'}}"
+    return messageText.replace(/\[(['"])(.*)\1\]/, `{{get ... $1$2$1}}`);
   }
 }
 

--- a/test-packages/test-utils/src/project.ts
+++ b/test-packages/test-utils/src/project.ts
@@ -122,10 +122,15 @@ export class Project {
     return project;
   }
 
-  public setGlintConfig(config: unknown): void {
+  public updateTsconfig(update: (config: TsconfigWithGlint) => TsconfigWithGlint | void): void {
     let tsconfig = JSON.parse(this.read('tsconfig.json'));
-    tsconfig.glint = config;
-    this.write('tsconfig.json', JSON.stringify(tsconfig, null, 2));
+    this.write('tsconfig.json', JSON.stringify(update(tsconfig) ?? tsconfig, null, 2));
+  }
+
+  public setGlintConfig(config: TsconfigWithGlint['glint']): void {
+    this.updateTsconfig((tsconfig) => {
+      tsconfig.glint = config;
+    });
   }
 
   public write(files: Record<string, string>): void;

